design/one-pager-ignore-changes.md

```diff
@@ -0,0 +1,354 @@
+  `desiredSize` due to resource requirement but crossplane decrease it to
+  `desiredSize` set in composition.
+- GCP [Node Pool], where when `management.autoUpgrade` is enabled. After
+  creation, the provider late initializes `spec.forProvider.version`,
+  which when the nodes get upgraded starts conflicting with the 
+  external nodes version. This causes an endless loop of downgrades
+  and upgrades.
+- Tags are a use case that affects multiple resources. Many cloud users
+  have external systems that add tags to resources for billing, tracking
+  or similar purposes. This can conflict with the tags set in 
+  `spec.forProvider.tags`.
+- There is an use case with [AzureAD Group] where initially the group is
+  created certain with `members` and `owners` but is then managed externally
+  over time as members need to be added or removed. For now the
+  workaround is to use terraform provider with [ignore_changes].
+
+In these cases, it would make sense to make Crossplane ignore select
+parameters during updates. Terraform addresses this use case with its
+[ignore_changes] functionality.
+
+An additional benefit of ignoring some fields would be that it would be
+possible to have different Crossplane Compositions managing a single
+external resource. This could open up some interesting new possibilities
+for Crossplane.
+
+The obvious way would be to just not set the field in the `spec.forProvider`.
+However, due to [Late Initialization], some fields are being set in
+`spec.forProvider` without any user input. Plus there are some cases
+where the field is required/wanted upon creation, but not on updates.
+
+So in this document, the aim will be to introduce a solution in Crossplane
+for ignoring some managed resource parameters during updates.
+
+## Goals
+
+* Design a functionality that allows for ignoring select fields during
+  resource updates.
+
+## Proposal
+
+Proposed solution is to use the new [managementPolicy] field which came with
+the [ObserveOnly] feature and add a new management policy that will 
+skip late initialization. The loss the information that the
+late initialization was providing would be offset by the `status.atProvider`
+which contains the resource state on the provider side and is being
+added alongside [ObserveOnly].
+
+In addition to the new management policy, we will also add a new field
+`spec.initProviders` which would contain parameters that should only be
+used at the creation time. This would allow users to specify fields that
+should be ignored on updates, but still be used during creation. This can
+be added as a separate feature.
+
+## Implementation
+
+### Management policy
+
+Proposed name for the new management policy is `PartialControl`, which
+would fit with the current `FullControl` and insinuate that the resource
+is not fully managed by Crossplane. Some other proposals for the naming
+are: `ExplicitControl`, `SelectiveControl`, `IgnoreLateInit`.
```

Comment:
Throwing another option into the ring: `SharedControl`.

I like this option because it communicates the _why_ a little better. Typically when Crossplane does not have "full" control of a resource it's because it's sharing control of the resource with some other system or entity.

Reply:
It's interesting to me that we're looking at several enum values that mostly end in the word _control_. This makes me wonder whether `managementPolicy` should instead be `controlPolicy`. For example:

```yaml
spec:
  controlPolicy: Full
```

Perhaps in this world `managementPolicy: ObserveOnly` would be `controlPolicy: None`. As in Crossplane does not control this resource. This would leave us with:

* `controlPolicy: Full` - The default. Crossplane controls every field, even those that are late-initialized.
* `controlPolicy: Shared` - Crossplane only controls fields that were explicitly specified.
* `controlPolicy: None` - Crossplane does not control this resource. It only observes it.

cc @turkenh

Reply:
>Throwing another option into the ring: `SharedControl`.

I feel like the `SharedControl` causes a slight confusion by triggering the question of "shares control with whom/what?". From Crossplane POV, it `partially` or `selectively` controls the resource, so I am closer to `Partial` or `Selective` control.

> It's interesting to me that we're looking at several enum values that mostly end in the word control. This makes me wonder whether managementPolicy should instead be controlPolicy.

I like the simplification in general, but not sure because of the following:

- `managementPolicy: ObserveOnly` is more explicit than `controlPolicy: None`.
- `managementPolicy` sounds easier to extend if needed compared to `controlPolicy`.

Reply:
> `managementPolicy` sounds easier to extend if needed compared to `controlPolicy`.

The `managementPolicy` field was heavily inspired by the previous work for [enhanced provider-kubernetes](https://github.com/crossplane-contrib/provider-kubernetes/blob/main/docs/enhanced-provider-k8s.md), where management policy was controlling which of the CRUD operations should be enabled for a given resource. The policy there [supports](https://github.com/crossplane-contrib/provider-kubernetes/blob/main/docs/enhanced-provider-k8s.md#resource-management-policy) `ObserveDelete` in addition to what we have here and there is a [request](https://github.com/crossplane-contrib/provider-kubernetes/issues/115) for `ObserveUpdate`. 

Noticed that we are moving away from such a semantic with `PartialControl` and by keeping `deletionPolicy` together with this. This should be ok since I am not aware of any current need for behaviors like `ObserveDelete` or `ObserveDelete` for cloud resources in general, but still, we may want to keep this in mind as possible extensions.

Reply:
my two cents, from a similar topic in a different context: thinking in policies is very procedural. You have to think about what the providers do, how they work. In that other context, we came up with ownership of fields as an alternative view, plus exceptions (a list of json paths), i.e.:
```
ownership: External
exceptFields:
- foo.bar
```
or:
```
ownership: Crossplane
externalFields:
- a.b.c
```
Looking further down, `ignoreChanges:` is pretty much that. `externallyOwned:` would reflect the ownership idea more, but that's cosmetics.

Reply:
> my two cents, from a similar topic in a different context: thinking in policies is very procedural. You have to think about what the providers do, how they work. In that other context, we came up with ownership of fields as an alternative view, plus exceptions (a list of json paths), i.e.:
> 
> ```
> ownership: External
> exceptFields:
> - foo.bar
> ```
> 
> or:
> 
> ```
> ownership: Crossplane
> externalFields:
> - a.b.c
> ```
> 
> Looking further down, `ignoreChanges:` is pretty much that. `externallyOwned:` would reflect the ownership idea more, but that's cosmetics.

Thanks for the feedback @sttts. Actually initially my proposed solution was to go with the `ignoreChanges:`, which is a valid idea and would solve the issue. But in the end, we opted for `initProviders` as it seems easier to grasp the concept to have an `initProviders` field that supplements the `forProviders` one at the creation/initialization time and later on the `forProviders` is the single source of truth. While with `ignoreChanges`, the field values that are in `forProviders` can become outdated if they are ignored, which can lead to confusion.